### PR TITLE
RHOL-1028: Clean up SortedParMap

### PR DESCRIPTION
**Pull request by tymm**
## Overview

- In a previous PR I changed `sortedMap` to `Map[Par,Par]` which caused a bug since `Map` does not preserve insertion order. I added a test case where the non-preservation of `Map` would show. 
- I added a generative test from @ArturGajowy that checks if sortedness is preserved.

I guess I should still extend from `IterableLike` but I thought I first put this out and see what @ArturGajowy says.

### Which JIRA issue does this PR relate to? If there is not a JIRA issue addressing this work, please create one now and add the link here.
https://rchain.atlassian.net/browse/RHOL-1028

### Complete this checklist before you submit the PR
- [x] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [x] If this PR adds a new feature, this PR includes tests related to this feature.
- [x] You assigned one person to review this PR

### CODE REVIEW

File: models/src/main/scala/coop/rchain/models/SortedParMap.scala [hunk 1]
```diff
@@ -1,29 +1,50 @@
...
+    case _                 => false
+  }
 }
 
 object SortedParMap {
   def apply(map: Map[Par, Par]): SortedParMap = new SortedParMap(map)
 
   def apply(seq: Seq[(Par, Par)]): SortedParMap = SortedParMap(seq.toMap)
 
+  def apply(sortedParMap: SortedParMap): SortedParMap = SortedParMap(sortedParMap.sortedMap)
```
Comment (ArturGajowy) [inline models/src/main/scala/coop/rchain/models/SortedParMap.scala]: Why is that needed? Why wouldnt `def apply(sortedParMap: SortedParMap): SortedParMap = sortedMap` suffice - and in that case - why even call the method?

File: models/src/main/scala/coop/rchain/models/SortedParMap.scala [hunk 2]
```diff
@@ -1,29 +1,50 @@
...
 import coop.rchain.models.rholang.sorter.ordering._
 
-class SortedParMap private (private val ps: Map[Par, Par])
-    extends Map[Par, Par]
-    with MapLike[Par, Par, SortedParMap] {
+import scala.collection.GenTraversableOnce
+import scala.collection.immutable.ListMap
 
-  lazy val sortedMap: Map[Par, Par] = ps.sort.toMap
+final class SortedParMap private (ps: Map[Par, Par]) extends Iterable[(Par, Par)] {
```
Comment (ArturGajowy) [inline models/src/main/scala/coop/rchain/models/SortedParMap.scala]: Why do you say we should implement IterableLike instead of Iterable?

**Comment by ArturGajowy** [comment]
I've restarted the build, but it failed again with (amongst others) a timeout in `StackSafetySpec`. Can you bump the timeout to 20 s?

File: models/src/main/scala/coop/rchain/models/SortedParMap.scala (hunk 1 above)
Comment (tymm) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: There is some code, e.g. `Reduce.scala:1294` that depends on this. Should I refactor the code that depends on that?

File: models/src/main/scala/coop/rchain/models/SortedParMap.scala [hunk 3]
```diff
@@ -1,29 +1,54 @@
...
+  lazy val sortedMapLookup: Map[Par, Par] = ps.sort.toMap
 
-  override def get(key: Par): Option[Par] = sortedMap.get(key)
+  def +(kv: (Par, Par)): SortedParMap = SortedParMap(sortedMap + kv)
 
-  override def iterator: Iterator[(Par, Par)] = sortedMap.toIterator
+  def ++(kvs: GenTraversableOnce[(Par, Par)]): SortedParMap = SortedParMap(sortedMap ++ kvs)
 
-  override def +[V1 >: Par](kv: (Par, V1)): Map[Par, V1] = sortedMap + kv
+  def -(key: Par): SortedParMap = SortedParMap(sortedMap - key)
```
Comment (ArturGajowy) [inline models/src/main/scala/coop/rchain/models/SortedParMap.scala]: the key itself may be not sorted. Take e.g. 

//ks == Sortable[Par].sort(ku), ks != ku
val m = SortedMap(ku -> 1)
val m2 = m - ku
//m2 == m, while it shouldn't
//this is because:
//m ~= Map(ks -> 1)

File: models/src/main/scala/coop/rchain/models/SortedParMap.scala [hunk 4]
```diff
@@ -1,29 +1,54 @@
...
 
-  override def -(key: Par): SortedParMap = SortedParMap(sortedMap - key)
+  def --(keys: GenTraversableOnce[Par]): SortedParMap = SortedParMap(sortedMap -- keys)
```
Comment (ArturGajowy) [inline models/src/main/scala/coop/rchain/models/SortedParMap.scala]: ditto

File: models/src/main/scala/coop/rchain/models/SortedParMap.scala [hunk 5]
```diff
@@ -1,29 +1,54 @@
...
+
+  def apply(par: Par): Par = sortedMapLookup(par)
```
Comment (ArturGajowy) [inline models/src/main/scala/coop/rchain/models/SortedParMap.scala]: same thing about sortedness of the input par

File: models/src/main/scala/coop/rchain/models/SortedParMap.scala [hunk 6]
```diff
@@ -1,29 +1,54 @@
...
+
+  def contains(par: Par): Boolean = sortedMapLookup.contains(par)
```
Comment (ArturGajowy) [inline models/src/main/scala/coop/rchain/models/SortedParMap.scala]: ditto

File: models/src/main/scala/coop/rchain/models/SortedParMap.scala [hunk 7]
```diff
@@ -1,29 +1,54 @@
...
+
+  def empty: SortedParMap = SortedParMap(Map.empty[Par, Par])
+
+  def get(key: Par): Option[Par] = sortedMapLookup.get(key)
```
Comment (ArturGajowy) [inline models/src/main/scala/coop/rchain/models/SortedParMap.scala]: ditto

File: models/src/main/scala/coop/rchain/models/SortedParMap.scala [hunk 8]
```diff
@@ -1,29 +1,54 @@
...
+
+  def getOrElse(key: Par, default: Par): Par = sortedMapLookup.getOrElse(key, default)
```
Comment (ArturGajowy) [inline models/src/main/scala/coop/rchain/models/SortedParMap.scala]: ditto

File: models/src/main/scala/coop/rchain/models/SortedParMap.scala (hunk 1 above)
Comment (ArturGajowy) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: How about this?

File: models/src/test/scala/coop/rchain/models/SortedParMapSpec.scala [hunk 9]
```diff
@@ -80,4 +83,73 @@ class SortedParMapSpec extends FlatSpec with Matchers {
...
+
+  it should "preserve sortedness of all elements and the whole map for all its operations" in {
+    import coop.rchain.models.testImplicits._
+
+    forAll { (pars1: Seq[Par], pars2: Seq[Par]) =>
+      val map1 = SortedParMap(pars1.zip(pars2))
+      val map2 = SortedParMap(pars2.zip(pars1))
+      checkSorted(map1)
+      checkSorted(map2)
+      Seq(map1.headOption, map2.headOption).flatten.foreach(kv => {
```
Comment (ArturGajowy) [inline models/src/test/scala/coop/rchain/models/SortedParMapSpec.scala]: Try to make this test find all the 'inputs have to be sorted' bugs. I think making this say
`Seq(parts1.headOption, parts2.headOption)` should suffice for the methods already called (please call the remaining ones as needed). If not, please find a different way to test that.

File: models/src/main/scala/coop/rchain/models/SortedParMap.scala [hunk 10]
```diff
@@ -1,29 +1,54 @@
...
+import scala.collection.GenTraversableOnce
+import scala.collection.immutable.ListMap
 
-  lazy val sortedMap: Map[Par, Par] = ps.sort.toMap
+final class SortedParMap private (ps: Map[Par, Par]) extends Iterable[(Par, Par)] {
 
-  override def empty: SortedParMap = SortedParMap(Map.empty[Par, Par])
+  // TODO: Merge `sortedMap` and `sortedMapLookup` into one VectorMap once available
+  lazy val sortedMap: ListMap[Par, Par]   = ListMap(ps.sort: _*)
+  lazy val sortedMapLookup: Map[Par, Par] = ps.sort.toMap
```
Comment (ArturGajowy) [inline models/src/main/scala/coop/rchain/models/SortedParMap.scala]: This means sorting twice. There's no need for two maps - just a:
 - List[Par] (for the key order) or List[(Par, Par)] (for key and entry order) 
and 
 - a HashMap[Par, Par] for the lookup
should suffice.

File: models/src/main/scala/coop/rchain/models/SortedParMap.scala (hunk 1 above)
Comment (ArturGajowy) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: Yes, please do. It's redundant.

File: models/src/main/scala/coop/rchain/models/SortedParMap.scala (hunk 3 above)
Comment (goral09) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: It should be enough to only sort the keys and not values when adding and removing elements from the Map and sort everything else when checking equality and before serialization.

Comment (ArturGajowy) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: Nopes, you'd also have to values when accessing elements. Also, since the types don't allow us to discern sorted from non-sroted ones, I want to keep the invariant of 'everything inside is sorted'.

Comment (goral09) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: Then sort when `EMap` are reduced/interpreted/evaled.

Comment (goral09) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: I am not sure if your approach is correct. You make internals leak to the user.
```scala
sortedMap.add(key, value).get(key) != value
```
because of sortedness. It is your call though.

Comment (ArturGajowy) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: Thanks for that remark! Indeed, that's undesirable.

Again though, the types are failing us here, because we've failed to use them in the first place. It is not an implementation detail of the SortedMap. Quite the contrary, it's the `raison d'etre` for it ;) SortedMap is used to represent a sorted map term. Such term has a unique canonical shape and this shape should be used here :)

Previously, the SortedMap had broken equals because `get` returned unsorted elements. But `SortedMap(1-> [2, 3])` should equal `SortedMap(1 -> [3, 2])`. Otherwise the latter map does not represent a sorted term :)

Let me know if that convinces you to the changed behavior :)

re: your proposal about postponing the sorting - I'd rather not reason about subsequent potential usages of the SortedMap and instead make all of them behave consistently. I'm not good at such reasoning. The types are great at that - but we haven't invited them :hits_blunt:

Comment (goral09) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: > But `SortedMap(1-> [2, 3])` should equal `SortedMap(1 -> [3, 2])`

Well, this example is not correct because `[2,3]` is not equal to `[3,2]` but `SortedMap(1 -> p | q)` is equal to `SortedMap(1 -> q | p)`. 
It also depends which algorithm is used for equality check. For example, if you use Rholang's `==` then `p == q != q == p`. 

Anyway, I think that users should not observe the sortedness property of a Map. The sorting was added for the serialization sake and has now bubbled up up to the user.

IMO there should be an internal, sorted representation for purposes of `contains` and `equals` but `get` should return unsorted value.

Comment (ArturGajowy) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: > if you use Rholang's == then `p == q != q == p`

WDYM? Rholang's `==` is not symmetric?

Also, I'd totally expect `{ p | q } == { q | p }` in rholang.

> sorting was added for the serialization sake 

AFAIK it's also a crucial concept behind all equality and pattern matching (pattern matching is just equality whenever the pattern has no free variables or wildcards).

Does that change your view of things?

Comment (tymm) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: I agree with @goral09 that `SortedMap(1-> [2, 3])` should not be equal to `SortedMap(1-> [3, 2])` since a List should keep its order.
AFAIK this is also how it is implemented right now.

Comment (ArturGajowy) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: I also agree with him on that :) I really meant the example he's given.

`SortedMap(1 -> q | p) == SortedMap(1 -> p | q)` should be true in current implementation, and in general.

Comment (ArturGajowy) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: > IMO there should be an internal, sorted representation for purposes of contains and equals but get should return unsorted value.

If this was true, the following would be false:

```
Map(1 -> 3 | 2).get(1) == 3 | 2
```

That is, if I'm able to create an unsorted (sub)term in the first place.
The thing is, at times the terms are being made by Scala code other than the normalizer (reductions, substitutions, `Method` calls) and there's no easy way to guarantee their outputs are sorted.

Since all the terms that can be written are 'born' sorted, there's no error in only returning sorted terms from the map.

Since the dynamically created terms (e.g. method call results) can't be guaranteed to be sorted, sorting them on entry might be costly, but is safer.

Comment (goral09) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: That's what I said. Rholang's `==` (equality sign) doesn't support these properties (like spatial matcher does). It just uses Java equality and I think this is something users should be aware of.

Comment (ArturGajowy) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: Since there's no way to have unsorted terms written in soruce code, and all rholang's internals should be returning sorted terms, there's nothing for the users to be aware of. 

*Throws chair.*

Comment (goral09) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: > Since there's no way to have unsorted terms written in soruce code,

That's not true. IIRC sorting happens only after substitution. It means that if you write down concrete, unsorted term then it stays unsorted.

Comment (ArturGajowy) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: Nopes, buildNormalizedTerm also sorts (see buildPar) :) I think we can finally agree :)

Comment (goral09) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: Doesn't it sort only the outermost elements? Not the inner ones? Like elements of the collection for example.

Comment (ArturGajowy) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: Nopes, sorting is recursive.

Comment (goral09) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: Then why do we have to sort at all when interacting with Map if everything is already sorted?

**Comment by ArturGajowy** [comment]
As I said, we don't have such guarantees when getting dynamically-generated
terms (subtitutions, Method results). I'd like the map to sort again until
we find a way to encode the sortedness in types or otherwise guarantee
sortedness of dynamically generated terms. Unless we prove that this is a
huge performance drag (i.e. have an e2e benchmark for that).

File: models/src/main/scala/coop/rchain/models/SortedParMap.scala (hunk 3 above)
Comment (ArturGajowy) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: As I said, we don't have such guarantees when getting dynamically-generated
terms (subtitutions, Method results). I'd like the map to sort again until
we find a way to encode the sortedness in types or otherwise guarantee
sortedness of dynamically generated terms. Unless we prove that this is a
huge performance drag (i.e. have an e2e benchmark for that).

File: models/src/main/scala/coop/rchain/models/SortedParMap.scala (hunk 10 above)
Comment (tymm) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: Done.

File: models/src/test/scala/coop/rchain/models/SortedParMapSpec.scala (hunk 9 above)
Comment (tymm) [reply models/src/test/scala/coop/rchain/models/SortedParMapSpec.scala]: Done.

File: models/src/main/scala/coop/rchain/models/SortedParMap.scala (hunk 8 above)
Comment (tymm) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: Done.

File: models/src/main/scala/coop/rchain/models/SortedParMap.scala (hunk 7 above)
Comment (tymm) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: Done.

File: models/src/main/scala/coop/rchain/models/SortedParMap.scala (hunk 6 above)
Comment (tymm) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: Done.

File: models/src/main/scala/coop/rchain/models/SortedParMap.scala (hunk 5 above)
Comment (tymm) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: Done.

File: models/src/main/scala/coop/rchain/models/SortedParMap.scala (hunk 4 above)
Comment (tymm) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: Done.

File: models/src/main/scala/coop/rchain/models/SortedParMap.scala (hunk 1 above)
Comment (tymm) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: Done.

File: models/src/main/scala/coop/rchain/models/SortedParMap.scala (hunk 2 above)
Comment (tymm) [reply models/src/main/scala/coop/rchain/models/SortedParMap.scala]: Not instead but in addition to. However seems like it is not needed for our purposes.

File: models/src/main/scala/coop/rchain/models/rholang/sorter/ordering.scala [hunk 11]
```diff
@@ -6,12 +6,24 @@ import coop.rchain.models.rholang.sorter.ScoredTerm._
...
+  implicit class ParOps(par: Par) {
+    implicit val sync = implicitly[Sync[Coeval]]
+
+    def sort: Par = Sortable[Par].sortMatch[Coeval](par).map(_.term).value()
+  }
+
+  implicit class GenTraversableOnceOps(ps: GenTraversableOnce[Par]) {
+    implicit val sync = implicitly[Sync[Coeval]]
+
+    def sort: GenTraversableOnce[Par] = ps.toList.sort
```
Comment (ArturGajowy) [inline models/src/main/scala/coop/rchain/models/rholang/sorter/ordering.scala]: Please remove those implicit conversions - I'd like Sortable to be the entry point for sorting for all new usages, as it's explicit about monads being used underneath. At some point we're going to have to join all the computations into a single one, in a single monad - otherwise performance will continue to suffer and there's no stacksafety

Comment (ArturGajowy) [reply models/src/main/scala/coop/rchain/models/rholang/sorter/ordering.scala]: parOps.sort should be just sort(Par) - a private method in the SortedMap.

GenTraverasableOps is not needed at all - it's enough if we just:
`kvs.foldLeft(sortedMap) { map, kv -> sortedMap - kv }` in the `--` mehtod
